src/main/java/org/betonquest/betonquest/api/profiles/ProfileKeyMap.java

```diff
@@ -0,0 +1,139 @@
+package org.betonquest.betonquest.api.profiles;
+
+import org.betonquest.betonquest.utils.PlayerConverter;
+import org.bukkit.Server;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to cache profiles.
+ * <p>
+ * It implements a simple cache that stores the profile's UUIDs as key in a map.
+ * It uses T as value.
+ * The interface runs via profiles.
+ *
+ * @param <T> the type of the object to be cached
+ */
+public class ProfileKeyMap<T> implements Map<Profile, T> {
+
+    /**
+     * The map used to store the profile's UUIDs.
+     */
+    private final Map<UUID, T> profileMap;
+
+    /**
+     * The server instance.
+     */
+    private final Server server;
+
+    /**
+     * Creates a new profile cache.
+     *
+     * @param server the server instance
+     */
+    public ProfileKeyMap(final Server server) {
+        this.profileMap = new HashMap<>();
+        this.server = server;
+    }
+
+    @Override
+    public int size() {
+        return profileMap.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return profileMap.isEmpty();
+    }
+
+    @Override
+    public boolean containsKey(final Object key) {
+        if (key instanceof Profile profile) {
+            return profileMap.containsKey(profile.getProfileUUID());
+        }
+        return false;
+    }
+
+    @Override
+    public boolean containsValue(final Object value) {
+        return profileMap.containsValue(value);
+    }
+
+    @Override
+    public T get(final Object key) {
+        if (key instanceof Profile profile) {
+            return profileMap.get(profile.getProfileUUID());
+        }
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public T put(final Profile key, final T value) {
+        return profileMap.put(key.getProfileUUID(), value);
+    }
+
+    @Override
+    public T remove(final Object key) {
+        if (key instanceof Profile profile) {
+            return profileMap.remove(profile.getProfileUUID());
+        }
+        return null;
+    }
+
+    @Override
+    public void putAll(@NotNull final Map<? extends Profile, ? extends T> map) {
+        final Map<UUID, ? extends T> entries = map.entrySet().stream()
+                .collect(Collectors.toMap(entry -> entry.getKey().getProfileUUID(), Entry::getValue));
+        profileMap.putAll(entries);
+    }
+
+    @Override
+    public void clear() {
+        profileMap.clear();
+    }
+
+    @NotNull
+    @Override
+    public Set<Profile> keySet() {
+        return profileMap.keySet().stream()
+                .map(uuid -> PlayerConverter.getID(server.getOfflinePlayer(uuid)))
+                .collect(Collectors.toSet());
```

Comment:
This implementation violates the interface definition: https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/Map.html#keySet()

The returned `Set` must be linked to the map, so `remove()` calls on the map (or it's iterator) must also remove the entries from this map.